PyVRP/PyVRP
# Support latest start on vehicle type

This PR:
Adds support for constraining start times for routes.

- [x] Closes #639  .
- [x] Adds and passes relevant tests.
- [x] Has well-formatted code and documentation.

Implementation details:
- Renamed `twEarly` to `earliestStart` and `twLate` to `latestFinish` in `VehicleType`
- Introduced attribute `latestStart` to `VehicleType`
- Not enforced that `latestStart <= latestFinish`. If `latestStart > latestFinish`, then it is possible to start the route after the latest finish (but this will always result in time warp). 
- Distinction between creating a `DurationSegment` for the start depot or end depot of the `VehicleType`. When creating a duration segment for start depot, then `vehicleType.latestStart` is used for `twLate` of the `DurationSegment`. For the end depot, `vehicleType.latestFinish` is used for `twLate`.

Open questions:
1. Renaming `twEarly` and `twLate` on `VehicleType` is a breaking change, but I do prefer to update the names to `earliestStart` and `latestFinish` for clarity. Should the attributes `twEarly` and `twLate` on `Client` also be renamed for consistency and to make clear that the service should start in this time window but not necessarily end in this window?

2. What should be the default value for `latestStart` in the constructor of `VehicleType`? The most logical default value would be "infinite", just like for `latestFinish` when it is not specified. However, the question is what to do when a value is provided for `latestFinish` but not for `latestStart`. Consider the following options:
    - Set `latestStart` to default value 'infinite' => route can start after latest finish. This was previously not possible, so this leads to the following change in behavior:
        - New behavior: We set the `latestStart` to 'infinite' by default, then the route can start after the route latest finish time. This prevents waiting time to get to the first client (if the client has a time window starting after the latest finish).
        - Old behavior: route cannot start after the latest finish, therefore it is not possible to delay the route start to prevent waiting time at the first client. Solution with old behavior will have higher duration and equal time warp, so it is seen as worse compared to the new behavior.
        - Note that this change in behavior only occurs in unlikely situations where the routes are infeasible anyway. So likely the effect is negligible and if there is an effect, it would be improving the objective.
    - Set `latestStart` to same value as `latestFinish` => backwards compatible (`latestStart` does not have to be provided when `latestFinish` is provided). However, this does require the `latestStart` argument to become of type `Optional` with default `None` in the constructor, which is inconsistent with the other arguments. Option 1 is chosen in this PR for that reason.

3. Should we support `earliestFinish` as well on a `VehicleType`? This would make it fully general/flexible, but it is maybe less common to have a restriction on the earliest finish of a route?
    - Note that in the issue description (https://github.com/PyVRP/PyVRP/issues/639) it is proposed to remove `latestFinish` such that the latest end of the route would be defined by `latestStart + maxDuration`. However, I would argue that `latestFinish` and `maxDuration` are two different concepts, so I propose to keep `latestFinish`. For example. a vehicle belonging to a vehicle type might start between 7-9AM, has a latest finish of 4PM and a max duration of 8 hours. If the route assigned to this vehicle takes up 8 hours (max duration not violated), then the route must start no later than 8AM to finish in time. If we would not consider `latestFinish`, then the route could have started at 9AM and ended at 5PM, while not violating the max duration.

<details>
</details>
